File: flowc/board/sync.py
"""Board-to-beads sync functionality."""

from __future__ import annotations

import re

from flowc.board.api import add_issue_to_board, list_board_items
from flowc.shared.config import load_beads

GITHUB_REF_PATTERN = re.compile(r"GitHub:\s*([^#\s]+)#(\d+)")


def get_p0_beads_with_github_refs() -> list[dict]:
    """Get all P0 beads that have GitHub issue references.

    Returns list of dicts with keys: bead_id, title, repo, issue_number.
    """
    beads = load_beads()
    p0_beads = []

    for bead in beads:
        # Check if P0 (priority stored as integer field)
        if bead.get("priority") != 0:
            continue

        # Extract GitHub reference from description
        desc = bead.get("description", "")
        github_match = GITHUB_REF_PATTERN.search(desc)
        if not github_match:
            continue

        p0_beads.append(
            {
                "bead_id": bead.get("id", ""),
                "title": bead.get("title", ""),
                "repo": github_match.group(1),
                "issue_number": int(github_match.group(2)),
            }
        )

    return p0_beads


def get_board_issues(board_key: str) -> list[dict]:
    """Get all issues on a board.

    Returns list of dicts with keys: item_id, title, repo, issue_number, status.
    """
    items = list_board_items(board_key)
    issues = []

    for item in items:
        content = item.get("content", {})
        if content.get("type") != "Issue":
            continue

        repo = content.get("repository", "")
        number = content.get("number")
        if not repo or not number:
            continue

        issues.append(
            {
                "item_id": item.get("id"),
                "title": item.get("title", ""),
                "repo": repo,
                "issue_number": number,
                "status": item.get("status", ""),
            }
        )

    return issues


def sync_board_with_beads(board_key: str, fix: bool = False) -> dict:
    """Compare P0 beads with board items and report/fix mismatches.

    Args:
        board_key: Board to sync (e.g., 'matsengrp/30').
        fix: If True, automatically fix mismatches.

    Returns:
        Dict with keys: missing_from_board, not_in_beads, fixes_applied.
    """
    p0_beads = get_p0_beads_with_github_refs()
    board_issues = get_board_issues(board_key)

    # Build sets for comparison
    p0_refs = {(b["repo"], b["issue_number"]) for b in p0_beads}
    board_refs = {(i["repo"], i["issue_number"]) for i in board_issues}

    # P0 beads not on board
    missing_from_board = []
    for bead in p0_beads:
        ref = (bead["repo"], bead["issue_number"])
        if ref not in board_refs:
            missing_from_board.append(bead)

    # Board issues not in P0 beads
    not_in_beads = []
    for issue in board_issues:
        ref = (issue["repo"], issue["issue_number"])
        if ref not in p0_refs:
            not_in_beads.append(issue)

    fixes_applied = []

    if fix:
        # Add missing P0s to board
        for bead in missing_from_board:
            success, msg = add_issue_to_board(
                board_key,
                bead["issue_number"],
                repo=bead["repo"],
                status="next",  # Default to Next column
            )
            if success:
                fixes_applied.append(f"Added #{bead['issue_number']} to board")
            else:
                fixes_applied.append(f"Failed to add #{bead['issue_number']}: {msg}")

        # Remove non-P0s from board (optional, commented out for safety)
        # for issue in not_in_beads:
        #     success, msg = remove_issue_from_board(
        #         board_key, issue["issue_number"], repo=issue["repo"]
        #     )
        #     if success:
        #         fixes_applied.append(f"Removed #{issue['issue_number']} from board")

    return {
        "missing_from_board": missing_from_board,
        "not_in_beads": not_in_beads,
        "fixes_applied": fixes_applied,
    }


def print_sync_report(result: dict, board_key: str):
    """Print sync report to stdout."""
    missing = result["missing_from_board"]
    not_in_beads = result["not_in_beads"]
    fixes = result.get("fixes_applied", [])

    print(f"## Board Sync: {board_key}\n")

    if missing:
        print(f"**P0 beads not on board** ({len(missing)}):")
        for bead in missing:
            print(f"  - {bead['repo']}#{bead['issue_number']}: {bead['title'][:50]}")
            print(f"    Bead: {bead['bead_id']}")
        print()

    if not_in_beads:
        print(f"**Board issues without P0 bead** ({len(not_in_beads)}):")
        for issue in not_in_beads:
            print(f"  - {issue['repo']}#{issue['issue_number']}: {issue['title'][:50]}")
            print(f"    Status: {issue['status']}")
        print()

    if fixes:
        print("**Fixes applied:**")
        for fix in fixes:
            print(f"  - {fix}")
        print()

    if not missing and not not_in_beads:
        print("Board and P0 beads are in sync!")

File: internal/flow/board/sync.go
package board

import (
	"fmt"

	"github.com/matsen/bipartite/internal/flow"
)

// SyncResult contains the result of a board sync operation.
type SyncResult struct {
	MissingFromBoard []flow.BeadGitHubRef // P0 beads not on board
	NotInBeads       []flow.BoardItem     // Board issues without P0 bead
	FixesApplied     []string             // Messages about fixes applied
}

// SyncBoardWithBeads compares P0 beads with board items and optionally fixes mismatches.
func SyncBoardWithBeads(boardKey string, fix bool) (*SyncResult, error) {
	p0Beads, err := flow.GetP0BeadsWithGitHubRefs()
	if err != nil {
		return nil, fmt.Errorf("loading P0 beads: %w", err)
	}

	boardIssues, err := ListBoardItems(boardKey)
	if err != nil {
		return nil, fmt.Errorf("listing board items: %w", err)
	}

	// Build sets for comparison
	p0Refs := make(map[string]flow.BeadGitHubRef)
	for _, b := range p0Beads {
		key := fmt.Sprintf("%s#%d", b.Repo, b.IssueNumber)
		p0Refs[key] = b
	}

	boardRefs := make(map[string]flow.BoardItem)
	for _, i := range boardIssues {
		if i.Content.Type != "Issue" {
			continue
		}
		key := fmt.Sprintf("%s#%d", i.Content.Repository, i.Content.Number)
		boardRefs[key] = i
	}

	result := &SyncResult{}

	// Find P0 beads not on board
	for key, bead := range p0Refs {
		if _, ok := boardRefs[key]; !ok {
			result.MissingFromBoard = append(result.MissingFromBoard, bead)
		}
	}

	// Find board issues not in P0 beads
	for key, issue := range boardRefs {
		if _, ok := p0Refs[key]; !ok {
			result.NotInBeads = append(result.NotInBeads, issue)
		}
	}

	// Apply fixes if requested
	if fix {
		for _, bead := range result.MissingFromBoard {
			err := AddIssueToBoard(boardKey, bead.IssueNumber, bead.Repo, "next")
			if err != nil {
				result.FixesApplied = append(result.FixesApplied,
					fmt.Sprintf("Failed to add #%d: %v", bead.IssueNumber, err))
			} else {
				result.FixesApplied = append(result.FixesApplied,
					fmt.Sprintf("Added #%d to board", bead.IssueNumber))
			}
		}
	}

	return result, nil
}

// PrintSyncReport prints a sync report to stdout.
func PrintSyncReport(result *SyncResult, boardKey string) {
	fmt.Printf("## Board Sync: %s\n\n", boardKey)

	if len(result.MissingFromBoard) > 0 {
		fmt.Printf("**P0 beads not on board** (%d):\n", len(result.MissingFromBoard))
		for _, bead := range result.MissingFromBoard {
			title := bead.Title
			if len(title) > 50 {
				title = title[:50] + "..."
			}
			fmt.Printf("  - %s#%d: %s\n", bead.Repo, bead.IssueNumber, title)
			fmt.Printf("    Bead: %s\n", bead.BeadID)
		}
		fmt.Println()
	}

	if len(result.NotInBeads) > 0 {
		fmt.Printf("**Board issues without P0 bead** (%d):\n", len(result.NotInBeads))
		for _, issue := range result.NotInBeads {
			title := issue.Title
			if len(title) > 50 {
				title = title[:50] + "..."
			}
			fmt.Printf("  - %s#%d: %s\n", issue.Content.Repository, issue.Content.Number, title)
			fmt.Printf("    Status: %s\n", issue.Status)
		}
		fmt.Println()
	}

	if len(result.FixesApplied) > 0 {
		fmt.Println("**Fixes applied:**")
		for _, fix := range result.FixesApplied {
			fmt.Printf("  - %s\n", fix)
		}
		fmt.Println()
	}

	if len(result.MissingFromBoard) == 0 && len(result.NotInBeads) == 0 {
		fmt.Println("Board and P0 beads are in sync!")
	}
}
